README split, edits, examples (#346)
diffkemp/diffkemp: README.md

options in two versions of a project and to report any discovered semantic
differences.

See the [Publication](#publications-and-talks) section to learn more about
DiffKemp.

> [!WARNING]
> DiffKemp is incomplete in its nature, hence may provide false positive
results (i.e. claiming that functions are not equivalent even though they are).
This especially happens with complex refactorings.

## Installation

You can install DiffKemp:

- By building it manually [from source](docs/installation.md)
- From a prepared RPM package that can be installed from our
  [Copr repository](https://copr.fedorainfracloud.org/coprs/viktormalik/diffkemp/):

  ```sh
  # Enabling the DiffKemp repository
  dnf install -y dnf-plugins-core
  dnf copr enable -y viktormalik/diffkemp
  # Installing DiffKemp
  dnf install -y diffkemp
  ```

## Usage

> ![DiffKemp commands](docs/img/commands.svg)

Author: Thanks, I added the checkmark and cross. I am thinking if I should also add an arrow to the semantic comparison indicating that it is possible to specify custom change patterns for the comparison.

  on a more complex program, specifically on the musl C library. We will learn
  how to compare complex `make`-based projects and how to use the result viewer.
- [Linux kernel](docs/examples/linux_kernel.md): Example of using DiffKemp
  on the Linux kernel.

## Why to use DiffKemp?

- **Efficiency in code review**: DiffKemp can reduce the amount of code that you will need to review manually by eliminating changes that do not impact the behaviour of the program.
- **Detection of unintended side effects**: Comparing programs on a semantic level can detect subtle changes that might alter the program's behavior.
- **Focus on critical functions**: You can specify a list of functions which you want to compare.
- **Support for large and complex C projects** like Linux kernel, C standard libraries, cryptographic libraries, ...

## How does it work?

The main focus of DiffKemp is high scalability, such that it can be applied to
large-scale projects containing a lot of code. To achieve that, the analysed
functions are first compiled into LLVM IR, then several code transformations are
applied, and finally the comparison itself is performed.

> Wherever possible, DiffKemp tries to compare instruction-by-instruction (on LLVM

Reviewer: I'd use `at the level of LLVM IR` instead of `on LLVM IR instructions`

IR instructions) which is typically sufficient for most of the code. When not
sufficient, DiffKemp tries to apply one of the built-in or user-supplied

> *semantics-preserving patterns*. If a semantic difference is still discovered,

Reviewer: ```suggestion
*semantics-preserving patterns*. If no pattern can be applied,
```

  you will need to review manually by eliminating changes that do not impact
  the behaviour of the program.
- **Detection of unintended side effects**: Comparing programs on a semantic
  level can detect subtle changes that might alter the program's behavior.
- **Focus on critical functions**: You can specify a list of functions which
  you want to compare.
- **Support for large and complex C projects** like Linux kernel, C standard
  libraries, cryptographic libraries, ...

## How does it work?

The main focus of DiffKemp is high scalability, such that it can be applied to
large-scale projects containing a lot of code. To achieve that, the analysed
functions are first compiled into LLVM IR, then several code transformations are
applied, and finally the comparison itself is performed.

Wherever possible, DiffKemp tries to compare instruction-by-instruction (at the
level of LLVM IR) which is typically sufficient for most of the code. When not
sufficient, DiffKemp tries to apply one of the built-in or user-supplied
*semantics-preserving patterns*. If no pattern can be applied,
the relevant diffs are reported to the user.

## Development

- [Development guide](docs/development.md)

> ## Important implementation details

Author: I deleted the section and moved the debug info sentence to brackets in snapshot generation paragraph.

options in two versions of a project and to report any discovered semantic
differences.

> [!WARNING]
> DiffKemp is incomplete in its nature, hence may provide false positive
results (i.e. claiming that functions are not equivalent even though they are).
This especially happens with complex refactorings.

## Installation

You can install DiffKemp:

- By building it manually [from source](docs/installation.md)
- From a prepared RPM package that can be installed from our
  [Copr repository](https://copr.fedorainfracloud.org/coprs/viktormalik/diffkemp/):

  ```sh
  # Enabling the DiffKemp repository
  dnf install -y dnf-plugins-core
  dnf copr enable -y viktormalik/diffkemp
  # Installing DiffKemp
  dnf install -y diffkemp
  ```

## Usage

```mermaid
---
config:
  themeCSS: "
% redefinition of subgraphs labels
.cluster-label {
  font-family: monospace;
  font-size: 13px;
}
"
  themeVariables:
    nodeBorder: black
    mainBkg: white
    clusterBkg: white
    clusterBorder: white
---
flowchart LR
    %% nodes
    pv(project versions)
    sl(symbol list)

>     subgraph sgSubgraph[diffkemp build\ndiffkemp build-kernel\ndiffkemp llvm-to-snapshot]

Reviewer 2: Is there a way to make these labels in "code-style font", i.e. `diffkemp build`, ...? It would help to understand that they are commands.

Author: I changed the font to [monospace](https://github.com/PLukas2018/diffkemp/blob/4ae8ff05360d3fa9d353bdef93f589a316d9d082/README.md?plain=1#L41), it looks like it was not previously aplicated because of comment which I added, now it seems to be fine.

- Tomáš Vojnar
- Petr Šilling
- Pavol Žáčik
- František Nečas
- Tomáš Kučma
- Lukáš Petr
- Tatiana Malecová
- Jakub Rozek

## Publications and talks

There is a number of publications and talks related to DiffKemp:

> - NETYS'22 talk  

Reviewer: I'd suggest keeping NETYS at the bottom, even though it is the newest. Custom patterns are quite specific and advanced and I think it's better to have a general introduction (the ICST paper and devconf talk) at the top.

- Tomáš Vojnar
- Petr Šilling
- Pavol Žáčik
- František Nečas
- Tomáš Kučma
- Lukáš Petr
- Tatiana Malecová
- Jakub Rozek

## Publications and talks

There is a number of publications and talks related to DiffKemp:

- ICST'21 [paper](https://ieeexplore.ieee.org/document/9438578)
  and [talk](https://zenodo.org/record/4658966):  
  V. Malík and T. Vojnar, "Automatically Checking Semantic Equivalence between
  Versions of Large-Scale C Projects," 2021 14th IEEE Conference on Software
  Testing, Verification and Validation (ICST), 2021, pp. 329-339.
- DevConf.CZ'19 talk  

>   [![DevConf.CZ'19 talk](http://img.youtube.com/vi/PUZSaLf9exg/0.jpg)](https://www.youtube.com/watch?v=PUZSaLf9exg)

Author: I reverted it.
